File: client/src/test/java/ch/ergon/modern/java/client/shape/ShapeAlgorithmsTest.java
package ch.ergon.modern.java.client.shape;

import ch.ergon.modern.java.api.shape.Circle;
import ch.ergon.modern.java.api.shape.Rectangle;
import ch.ergon.modern.java.api.shape.ShapeFactory;
import ch.ergon.modern.java.api.shape.Square;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class ShapeAlgorithmsTest {

    private final ShapeFactory factory = ShapeFactory.getInstance();

    private final ShapeAlgorithms algorithms = new ShapeAlgorithms();

    @Test
    public void testCopySquare() {
        var original = factory.createSquare(10.0);
        var copy = assertInstanceOf(Square.class, algorithms.copy(original));

        assertNotSame(original, copy);
        assertEquals(original, copy);
        assertEquals(10.0, copy.getSide());
    }

    @Test
    public void testCopyCircle() {
        var original = factory.createCircle(5.0);
        var copy = assertInstanceOf(Circle.class, algorithms.copy(original));

        assertNotSame(original, copy);
        assertEquals(original, copy);
        assertEquals(5.0, copy.getRadius());
    }

    @Test
    public void testCopyRectangle() {
        var original = factory.createRectangle(1.0, 2.0);
        var copy = assertInstanceOf(Rectangle.class, algorithms.copy(original));

        assertNotSame(original, copy);
        assertEquals(original, copy);
        assertEquals(1.0, copy.getWidth());
        assertEquals(2.0, copy.getHeight());
    }

    @Test
    public void testSimplifySquare() {
        var original = factory.createSquare(3.0);
        var simplified = assertInstanceOf(Square.class, algorithms.simplify(original));

        assertSame(original, simplified);
    }

    @Test
    public void testSimplifyCircle() {
        var original = factory.createCircle(7.5);
        var simplified = assertInstanceOf(Circle.class, algorithms.simplify(original));

        assertSame(original, simplified);
    }

    @Test
    public void testSimplifySquareRectangle() {
        var original = factory.createRectangle(5.0, 5.0);
        var simplified = assertInstanceOf(Square.class, algorithms.simplify(original));

        assertEquals(5.0, simplified.getSide());
    }

    @Test
    public void testSimplifyNormalRectangle() {
        var original = factory.createRectangle(5.0, 10.0);
        var simplified = assertInstanceOf(Rectangle.class, algorithms.simplify(original));

        assertSame(original, simplified);
    }
}

File: client/src/test/kotlin/ch/ergon/modern/java/client/shape/kotlin/ShapeAlgorithmsTest.kt
package ch.ergon.modern.java.client.shape.kotlin;

import ch.ergon.modern.java.api.shape.kotlin.Circle;
import ch.ergon.modern.java.api.shape.kotlin.Rectangle;
import ch.ergon.modern.java.api.shape.kotlin.ShapeFactory;
import ch.ergon.modern.java.api.shape.kotlin.Square;
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test;

class ShapeAlgorithmsTest {

    private val factory = ShapeFactory.instance;

    @Test
    fun testCopySquare() {
        val original = factory.createSquare(10.0);
        val copy = assertInstanceOf(Square::class.java, original.copy())

        assertNotSame(original, copy);
        assertEquals(original, copy);
        assertEquals(10.0, copy.side);
    }

    @Test
    fun testCopyCircle() {
        val original = factory.createCircle(5.0);
        val copy = assertInstanceOf(Circle::class.java, original.copy())

        assertNotSame(original, copy);
        assertEquals(original, copy);
        assertEquals(5.0, copy.radius);
    }

    @Test
    fun testCopyRectangle() {
        val original = factory.createRectangle(1.0, 2.0);
        val copy = assertInstanceOf(Rectangle::class.java, original.copy())

        assertNotSame(original, copy);
        assertEquals(original, copy);
        assertEquals(1.0, copy.width);
        assertEquals(2.0, copy.height);
    }

    @Test
    fun testSimplifySquare() {
        val original = factory.createSquare(3.0);
        val simplified = assertInstanceOf(Square::class.java, original.simplify())

        assertSame(original, simplified);
    }

    @Test
    fun testSimplifyCircle() {
        val original = factory.createCircle(7.5);
        val simplified = assertInstanceOf(Circle::class.java, original.simplify())

        assertSame(original, simplified);
    }

    @Test
    fun testSimplifySquareRectangle() {
        val original = factory.createRectangle(5.0, 5.0);
        val simplified = assertInstanceOf(Square::class.java, original.simplify())

        assertEquals(5.0, simplified.side);
    }

    @Test
    fun testSimplifyNormalRectangle() {
        val original = factory.createRectangle(5.0, 10.0);
        val simplified = assertInstanceOf(Rectangle::class.java, original.simplify())

        assertSame(original, simplified);
    }
}
